### PR TITLE
complete support of variable substitution

### LINKED ISSUE #3434: support vscode variable substitution
List of all variables supported by VS Code: https://code.visualstudio.com/docs/editor/variables-reference

### PR DESCRIPTION
**Pull request by akosyakov**
#### What it does

- fix #3434 - complete support of vscode variable substitution
  - only [input variables for tasks](https://code.visualstudio.com/docs/editor/variables-reference#_input-variables) are left out, since it depends on #5013 and requires substantial refactoring of the task extension first
- fixed resolution of env variables, see https://github.com/theia-ide/theia/pull/5835/commits/fa3016e44c0b143530aa763dd217437c5f1f7e58
- [x] a CQ for copied input schema and types: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20504

#### How to test

- look at https://code.visualstudio.com/docs/editor/variables-reference and test different variables
- you can also follow commit history to see support for which variables were added
- for most variables, you can use `Variables: List All` command
  - then select a variable to test
  - then you will be prompted to provide an argument for a variable, an argument is a part of ':' in variable substitution expression
  - if a variable does not accept argument just press enter, if it accepts then provide a proper argument
  - you should see a variable value if it can be computed in the given context, or raw expression if it cannot
- alternatively, you can define variables in tasks.json and launch.json and then try to run a task or launch configuration
  - like defined [here](https://code.visualstudio.com/docs/editor/variables-reference#_why-isnt-workspaceroot-documented) for example, there is a bug though that an output for short-lived task is lost, you may want to add `watch echo` in order to keep a task running
- for [command variables](https://code.visualstudio.com/docs/editor/variables-reference#_command-variables) you may want to install vscode node js extensions instead of native and test with `extension.pickNodeProcess`, don't forget that you need both vscode nodejs extension to use them
- for input variables you can follow: https://code.visualstudio.com/docs/editor/variables-reference#_input-variables
  - tasks.json is not supported, since the task extension is not ready for it, see #5836

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully reviewed following [the review guidelines](https://github.com/theia-ide/theia/blob/ak/pr_template/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/ak/pr_template/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by kittaakos** [comment]
I am verifying it.

**Comment by kittaakos** [comment]
What should happen when a variable is not available from the shell?
For instance, when I am usign `"${env:PATH}"` then I can see value of the `PATH` variable is printed. If I use `"${env:USERNAME}"` it is printed as `${env:USERNAME}` instead of an empty string.

```
Akoss-MacBook-Pro:theia akos.kitta$ echo $USERNAME

Akoss-MacBook-Pro:theia akos.kitta$ 
```

File: packages/variable-resolver/src/browser/variable-quick-open-service.ts [hunk 1]
```diff
@@ -44,26 +63,15 @@ export class VariableQuickOpenService implements QuickOpenModel {
     onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): void {
         acceptor(this.items);
     }
-}
-
-export class VariableQuickOpenItem extends QuickOpenItem {
```
Comment (kittaakos) [inline packages/variable-resolver/src/browser/variable-quick-open-service.ts]: Should go to the breaking changes section.

**Comment by kittaakos** [comment]
[bug]: When defining a variable, select `Variable: List All` > `${env}` > <kbd>Esc</kbd>:

Screen Shot 2019-08-05 at 11 49 31

**Comment by akosyakov** [comment]
> If I use "${env:USERNAME}" it is printed as ${env:USERNAME} instead of an empty string

That's a great point! It should be an empty string for sure.

**Comment by kittaakos** [comment]
I have verified the followinngs I could understand from the description; they work.

Verified variables:
```
"${workspaceFolder}",
"${workspaceFolderBasename}",
"${file}",
"${relativeFile}",
"${fileBasename}",
"${fileBasenameNoExtension}",
"${fileDirname}",
"${fileExtname}",
"${cwd}",
"${lineNumber}",
"${selectedText}",
"${execPath}",
```

Executed module:
```js
console.log('ARGS:');
for (const e of process.argv.splice(2)) {
    console.log(e);
}
```

Running `./xxx/index.js`:
```
ARGS:
/Users/akos.kitta/git/theia
theia
/Users/akos.kitta/git/theia/xxx/index.js
xxx/index.js
index.js
index
/Users/akos.kitta/git/theia/xxx
.js
/Users/akos.kitta/git/theia
2
process.argv.splice(2)
/Users/akos.kitta/.nvm/versions/node/v10.15.3/bin/node
```

Running `./index.js`:
```
ARGS:
/Users/akos.kitta/git/theia
theia
/Users/akos.kitta/git/theia/index.js
index.js
index.js
index
/Users/akos.kitta/git/theia
.js
/Users/akos.kitta/git/theia
2
process.argv.splice(2)
/Users/akos.kitta/.nvm/versions/node/v10.15.3/bin/node
```

I **did no** verify the case sensitivity/insensitivity:
> Note: Be sure to match the environment variable name's casing, for example ${env:Path} on Windows.

**Comment by akosyakov** [comment]
> I did no verify the case sensitivity/insensitivity:

I've noticed that VS Code always convert env var name to lower case on windows: https://github.com/microsoft/vscode/blob/0a34756cae4fc67739e60c708b04637089f8bb0d/src/vs/workbench/services/configurationResolver/common/variableResolver.ts#L172-L175 Should we do the same?

File: packages/variable-resolver/src/browser/variable-quick-open-service.ts (hunk 1 above)
Comment (akosyakov) [reply packages/variable-resolver/src/browser/variable-quick-open-service.ts]: ok

**Comment by akosyakov** [comment]
> Unavailable variables should be empty strings

I don't think it is a case for all variables, just for env because they are evaluated to empty string when missing. In other cases it is not true. Maybe we should throw in other cases to prevent running tasks and launch configuration with bogus arguments.

**Comment by akosyakov** [comment]
I will wait till CQ is resolved before continuing working on it. It is pending already for 5 days and they don't seem to work on it: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20504#c3 cc @marcdumais-work

**Comment by marcdumais-work** [comment]
> It is pending already for 5 days and they don't seem to work on it

Yes, I noticed things seem slow at the Foundation recently; I suspect late summer vacations.

**Comment by vince-fugnitto** [comment]
> I will wait till CQ is resolved before continuing working on it. It is pending already for 5 days and they don't seem to work on it: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20504#c3 cc @marcdumais-work

I believe the [CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20504#c3) is waiting for some form of feedback from you @akosyakov :( I see the message _"IPTeam awaiting response from Committer."_

**Comment by marcdumais-work** [comment]
> I believe the [CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20504#c3) is waiting for some form of feedback from you @akosyakov :( I see the message _"IPTeam awaiting response from Committer."_

Thanks for pointing this out, Vince. I answered the question and sent the CQ back to the IP team.

**Comment by akosyakov** [comment]
@kittaakos I've pushed a commit to fix resolving env variables values on windows and evaluating them to an empty string if they are missing to align with the shell behaviour. I'm going to merge it tomorrow if no one objects and CI is green.

**Pull request merged**